src/ORM/Connect/TempDatabase.php

```diff
@@ -232,7 +245,12 @@ public function resetDBSchema(array $extraDataObjects = [])
         // pgsql doesn't allow schema updates inside transactions
         // so we need to rollback any transactions before commencing a schema reset
         if ($this->hasStarted()) {
-            $this->rollbackTransaction();
```

Comment:
Ok it turns out this code doesn't work.

Trying

```php
public function resetDBSchema(array $extraDataObjects = [])
    {
        if (!$this->isUsed()) {
            return;
        }
        try {
            $this->rebuildTables($extraDataObjects);
        } catch (DatabaseException $ex) {
            // In case of error during build force a hard reset
            // e.g. pgsql doesn't allow schema updates inside transactions
            $this->kill();
            $this->build();
            $this->rebuildTables($extraDataObjects);
        }
    }
```

@dhensby seem ok?

Reply:
```
PHP Fatal error:  Uncaught SilverStripe\ORM\Connect\DatabaseException: Couldn't run query:

SELECT tablename
FROM pg_catalog.pg_tables
WHERE schemaname = $1 AND tablename NOT ILIKE 'pg\\_%' AND tablename NOT ILIKE 'sql\\_%'

ERROR:  current transaction is aborted, commands ignored until end of transaction block in /Users/dmooyman/Sites/ss42test/vendor/silverstripe/framework/src/ORM/Connect/DBConnector.php:64
Stack trace:
#0 /Users/dmooyman/Sites/ss42test/vendor/silverstripe/postgresql/code/PostgreSQLConnector.php(234): SilverStripe\ORM\Connect\DBConnector->databaseError('Couldn't run qu...', 256, 'SELECT tablenam...', Array)
#1 /Users/dmooyman/Sites/ss42test/vendor/silverstripe/framework/src/ORM/Connect/Database.php(168): SilverStripe\PostgreSQL\PostgreSQLConnector->preparedQuery('SELECT tablenam...', Array, 256)
#2 /Users/dmooyman/Sites/ss42test/vendor/silverstripe/framework/src/ORM/Connect/Database.php(222): SilverStripe\ORM\Connect\Database->SilverStripe\ORM\Connect\{closure}('SELECT tablenam...')
#3 /Users/dmooyman/Sites/s in /Users/dmooyman/Sites/ss42test/vendor/silverstripe/framework/src/ORM/Connect/DBConnector.php on line 64
```

WHY

Reply:
I think we need rollbackRecursive() on database.

Reply:
It's doing the rollback, but the rollback doesn't seem to actually rollback on postgres.

Reply:
ok, there is like 6 levels of nesting at this error.

Reply:
I was dumb. I was doing a `BEGIN` followed IMMEDIATELY by a `ROLLBACK` somehow.

Reply:
ok the problem is hasStarted is true sometimes when there is no transaction. After a rollback hasStarted needs to stay true, even though the transaction is rolled back. It's a badly documented var.

Reply:
I'm dropping hasStarted() and using isUsed() instead. Between runs, there will be two states:

 - The DB Is populated, assume that it's ready and safe to use.
 - The DB is killed, assume that it requires a rebuild.

There is no "there but needs a reset state".